### PR TITLE
mgr/cephadm: try to avoid pull when getting container image info

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/58051
Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/serve.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/serve.py:1454]: It seems like the second check (the one running `pull`) could go into the except clause.

**Review by mgfritch — APPROVED** [review]

File: src/pybind/mgr/cephadm/serve.py
Comment (mgfritch) [reply src/pybind/mgr/cephadm/serve.py:1454]: maybe report the exception cause via `log.debug` ?

Comment (trociny) [reply src/pybind/mgr/cephadm/serve.py:1454]: It is rather expected situation that "inspect" may fail, and we have similar code in upgrade [1] and we don't have any debug loggin there. And we have enough debug logging in `_run_cephadm`, which will be called, so one would see the result. Still if you think a `log.debug` message would be useful here I don't mind adding it.

[1] https://github.com/ceph/ceph/blob/main/src/pybind/mgr/cephadm/upgrade.py#L836

**Comment by trociny** [comment]
I set DNM flag for this PR only because I have not tested it yet. Is there a teuthology suite I could run on sepia lab to test it?

File: src/pybind/mgr/cephadm/serve.py
Comment (mgfritch) [reply src/pybind/mgr/cephadm/serve.py:1454]: ah yep! the debug logging in `_run_cephadm` is sufficient, there really is no need for additional logging here.

**Comment by trociny** [comment]
jenkins test make check

**Comment by trociny** [comment]
Hm, the cephadm unit tests are failing and it looks like related to my change, but I have no idea what could be wrong. Does anyone with cephaadm unit tests experience have an idea?

The failing test:
```
_______________________ TestCephadm.test_apply_osd_save ________________________

self = <cephadm.tests.test_cephadm.TestCephadm object at 0x7fb6c54d8f40>
_run_cephadm = <AsyncMock name='_run_cephadm' id='140420875697696'>
cephadm_module = <cephadm.module.CephadmOrchestrator object at 0x7fb64829d9f0>

    @mock.patch("cephadm.serve.CephadmServe._run_cephadm")
    def test_apply_osd_save(self, _run_cephadm, cephadm_module: CephadmOrchestrator):
        _run_cephadm.side_effect = async_side_effect(('{}', '', 0))
        with with_host(cephadm_module, 'test'):
    
            spec = DriveGroupSpec(
                service_id='foo',
                placement=PlacementSpec(
                    host_pattern='*',
                ),
                data_devices=DeviceSelection(
                    all=True
                )
            )
    
            c = cephadm_module.apply([spec])
            assert wait(cephadm_module, c) == ['Scheduled osd.foo update...']
    
            inventory = Devices([
                Device(
                    '/dev/sdb',
                    available=True
                ),
            ])
    
            cephadm_module.cache.update_host_devices('test', inventory.devices)
    
            _run_cephadm.side_effect = async_side_effect((['{}'], '', 0))
    
            assert CephadmServe(cephadm_module)._apply_all_services() is False
    
            _run_cephadm.assert_any_call(
                'test', 'osd', 'ceph-volume',
                ['--config-json', '-', '--', 'lvm', 'batch',
                    '--no-auto', '/dev/sdb', '--yes', '--no-systemd'],
                env_vars=['CEPH_VOLUME_OSDSPEC_AFFINITY=foo'], error_ok=True, stdin='{"config": "", "keyring": ""}')
>           _run_cephadm.assert_any_call(
                'test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'], image='', no_fsid=False)

cephadm/tests/test_cephadm.py:794: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <AsyncMock name='_run_cephadm' id='140420875697696'>
args = ('test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'])
kwargs = {'image': '', 'no_fsid': False}
expected = call('test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'], image='', no_fsid=False)
cause = None
actual = [call('test', <CephadmNoImage.token: 1>, 'check-host', ['--expect-hostname', 'test'], addr='test', error_ok=True, no_f...ume', ['--', 'inventory', '--format=json-pretty', '--filter-for-batch'], no_fsid=False, error_ok=False, image=''), ...]
expected_string = "_run_cephadm('test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'], image='', no_fsid=False)"
```

So, `_run_cephadm test osd ceph-volume ...` was never called as expected, and it might be somehow related to my cephadm changes but I can't see how.

There is also this error during the test execution, which may be the real cause why the expected `_run_cephadm` call was not executed, but again I can't understand why it happens:
```
DEBUG    cephadm.serve:serve.py:521 _apply_all_services
DEBUG    cephadm.serve:serve.py:616 Applying service osd.foo spec
DEBUG    cephadm.services.osd:osd.py:32 Processing DriveGroup DriveGroupSpec.from_json(yaml.safe_load('''service_type: osd
service_id: foo
service_name: osd.foo
placement:
  host_pattern: '*'
spec:
  data_devices:
    all: true
  filter_logic: AND
  objectstore: bluestore
'''))
DEBUG    tests:mgr_module.py:1615 mon_command: 'osd tree' -> 0 in 0.000s
ERROR    cephadm.services.osd:osd.py:412 Cannot decode JSON: ''
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/services/osd.py", line 410, in refresh
    tree = json.loads(out)
  File "/usr/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by phlogistonjohn** [comment]
Hi,  I think what is happening is that every call to `_run_cephadm_json` now calls `_run_cephadm` with the `error_ok` argument. While the error_ok argument existed as valid for `_run_cephadm` it wasn't passed in the older version of  `_run_cephadm_json`   so that anything that uses mock to check what arguments _run_cephadm were called with will now see `error_ok` .

I think your options are to fix the test failures by finding all the sites that assert the arguments of _run_cephadm and update them. Example:

```diff
diff --git a/src/pybind/mgr/cephadm/tests/test_cephadm.py b/src/pybind/mgr/cephadm/tests/test_cephadm.py
index 76a1073fc4a..d0106b13dd4 100644
--- a/src/pybind/mgr/cephadm/tests/test_cephadm.py
+++ b/src/pybind/mgr/cephadm/tests/test_cephadm.py
@@ -792,9 +792,9 @@ class TestCephadm(object):
                     '--no-auto', '/dev/sdb', '--yes', '--no-systemd'],
                 env_vars=['CEPH_VOLUME_OSDSPEC_AFFINITY=foo'], error_ok=True, stdin='{"config": "", "keyring": ""}')
             _run_cephadm.assert_any_call(
-                'test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'], image='', no_fsid=False)
+                'test', 'osd', 'ceph-volume', ['--', 'lvm', 'list', '--format', 'json'], image='', no_fsid=False, error_ok=False)
             _run_cephadm.assert_any_call(
-                'test', 'osd', 'ceph-volume', ['--', 'raw', 'list', '--format', 'json'], image='', no_fsid=False)
+                'test', 'osd', 'ceph-volume', ['--', 'raw', 'list', '--format', 'json'], image='', no_fsid=False, error_ok=False)
 
     @mock.patch("cephadm.serve.CephadmServe._run_cephadm")
```

Or you need to conditionally pass the `error_ok` argument to `_run_cephadm` so that it doesn't change the code paths that where it was not previously used.  I think the former is the better option.

**Review by adk3798 — APPROVED** [review]
LGTM pending the tests fixes specified in https://github.com/ceph/ceph/pull/48973#issuecomment-1341157562

**Comment by trociny** [comment]
@phlogistonjohn Thank you! Looks simple after you explained :-)

Hope, it is fixed now.  The `make check`  still failed but this looks like it needs to wait for #49321 to be merged.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-01-21_05:38:21-orch:cephadm-wip-adk-testing-2023-01-20-1359-distro-default-smithi/

Lots of failures (13) but all accounted for

- 3 instances of `'list' object has no attribute 'get'`. Failure was on new test for setting mon crush location introduced by a PR in the run. That PR can't be merged, but this shouldn't block merging anything else.
- 2 instances of test_cephadm.sh test failing, caused by the same mon crush location PR. Failed with `local variable 'config_json' referenced before assignment`
- 1 instance of https://tracker.ceph.com/issues/49287, known issue
- 2 instances of https://tracker.ceph.com/issues/58524, known issue
- 2 instances of https://tracker.ceph.com/issues/58535, known issue
- 1 instance of https://tracker.ceph.com/issues/57771, known issue
- 1 instance of https://tracker.ceph.com/issues/57755, known issue
- 1 failed basic smoke test that timed out waiting for all 8 OSDs to come up. Did 5 reruns of this test with the same build (https://pulpito.ceph.com/adking-2023-01-22_15:28:20-orch:cephadm-wip-adk-testing-2023-01-20-1359-distro-default-smithi/) all of which passed. My guess is an I/O error creating the last OSD, which I've seen happen (infrequently) when we make a bunch of OSDs on the split up nvme drive we use for having multiple OSDs on these machines. Considering it a 1 off for now, and not blocking merging over it.

Overall, PRs in the run should be okay to merge other than the mon crush location one causing failures. Will start to try and clean up the test suite now that we're able to make builds and run tests again.

**Comment by adk3798** [comment]
@trociny it looks like there are still a couple test failures. I made a sample commit that fixes them on top of your changes https://github.com/adk3798/ceph/commit/b95c411491e74db97852d27eebac776e3bd4b331. With something like that to make make check pass, given the teuthology run was good, we'll be able to merge. Thanks, and sorry for the hassle on this.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-02-13_21:57:03-orch:cephadm-wip-adk-testing-2023-02-13-1359-distro-default-smithi/

11 failures, one dead job

- 8 of the failed jobs were on new tests being added in by other PRs in the run, which obviously aren't quite working yet (test_monitoring_stack_basic, rotate-ssh-keys, test_iscsi_container) but outside of iscsi and monitoring stack specific changes and the PRs that introduced the tests, this shouldn't block merging.
- 2 instances of https://tracker.ceph.com/issues/58535
- 1 instance of https://tracker.ceph.com/issues/49287
- 1 dead job was an error re-imaging machines, infra issue

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
In hindsight, this has broken being able to have your ceph image on a floating tag, and being able to upgrade to a new image pushed to that tag since it will just inspect the currently downloaded version and not pull and see the image now has a new digest. Maybe we should be trying the pull first and falling back to the inspect? I think that would cover this case while still allowing floating tags to work.

**Comment by trociny** [comment]
@adk3798 I think I am fine with changing to pull first and fallback to inspect. The behavior might still be confusing or unexpected for users though.

May be we just need to add a config option (`floatin_tag = false/true (default true)` or something like this)? I.e. we may leave the current behavior (try inspect first) but it will be used only if the option `floating_tag` is false. Or we may change to the proposed behaviour, i.e. always pull first, and fallback to inspect only if the `floating_tag` is false. I would vote for the first variant, t.e. leave the current behavior and just add the option.

**Comment by adk3798** [comment]
> @adk3798 I think I am fine with changing to pull first and fallback to inspect. The behavior might still be confusing or unexpected for users though.
> 
> May be we just need to add a config option (`floatin_tag = false/true (default true)` or something like this)? I.e. we may leave the current behavior (try inspect first) but it will be used only if the option `floating_tag` is false. Or we may change to the proposed behaviour, i.e. always pull first, and fallback to inspect only if the `floating_tag` is false. I would vote for the first variant, t.e. leave the current behavior and just add the option.

I sort of thought of the use_repo_digest option as already being the equivalent, since the only time using a repo digest over a tag is useful is when you have a floating tag. Doing the inspect first when that setting is turned off could be reasonable.

**Comment by trociny** [comment]
@adk3798 (re)using `use_repo_digest` option for this sounds good to me. Thanks. I will prepare a PR and we may have further discussion there.

**Comment by trociny** [comment]
> @adk3798 (re)using `use_repo_digest` option for this sounds good to me. Thanks. I will prepare a PR and we may have further discussion there.

#50311

**Pull request merged**
